packages/hadron-build/lib/signtool.js

```diff
@@ -1,22 +1,103 @@
 
-async function signtool(fileToSign) {
-  const signtoolPath = path.resolve(__dirname, '..', 'signtool/signtool.exe');
-
-  const execArgs = [signtoolPath, [path.resolve(fileToSign)], { stdio: 'inherit' }];
-  debug(`Running signtool.exe to sign '${signtoolPath}'`, {
-    execArgs: execArgs,
-    env: {
-      NOTARY_SIGNING_COMMENT: process.env.NOTARY_SIGNING_COMMENT,
-      NOTARY_URL: process.env.NOTARY_URL,
-      NOTARY_SIGNING_KEY: process.env.NOTARY_SIGNING_KEY,
-    }
-  });
+const canSign = () => (
+  process.env.GARASIGN_USERNAME &&
+  process.env.GARASIGN_PASSWORD &&
+  process.env.ARTIFACTORY_USERNAME &&
+  process.env.ARTIFACTORY_PASSWORD
+);
+
+/**
+ * When using gpg to sign a file, it creates a signature file
+ * with same name as the original file and adds `.sig` to it.
+ *
+ * @param {string} filename
+ * @returns string
+ */
+function getSignedFilename(filename) {
+  return `${filename}.sig`;
+}
+
+/**
+ * Currently, windows and macos zip are created from `zip` step
+ * of the release process and we sign them here. For linux, we
+ * create and sign the archive when creating corresponding deb/rpm file.
+ *
+ * @param {import('./Target')} target
+ */
+function signArchive(target, cb) {
+  const { app_archive_name, platform } = target;
+  if (platform === 'linux') {
+    debug('linux archive is signed when creating deb/rpm');
+    return cb();
+  }
+  sign(target.dest(app_archive_name)).then(cb).catch(cb);
+}
+
+/**
+ *
+ * For signing a file, we have following options with the current CI setup:
+ *
+ * |----------|--------|--------|----------------|--------------|
+ * | platform | docker | client | signing method | filetype     |
+ * |----------|--------|--------|----------------|--------------|
+ * | macOS    | no     | remote | gpg            | zip          |
+ * | ubuntu   | yes    | local  | gpg            | zip, deb     |
+ * | rhel     | yes ‡  | remote | gpg            | zip, rpm     |
+ * | windows  | no     | remote | jsign          | msi, exe     |
+ * | windows  | no     | remote | gpg            | zip, nupkg   |
+ * |----------|--------|--------|----------------|--------------|
+ *
+ * ‡ On rhel we have docker, but currently we are using rhel7.6
+ *  for packaging and docker does not support login using --password-stdin
+ *
+ * @param {string} src
+ * @returns {Promise<void>}
+ */
+async function sign(src, garasign = _garasign) {
+  const variant = process.env.EVERGREEN_BUILD_VARIANT;
+  debug('Signing on %s ... %s', variant, src);
 
-  // eslint-disable-next-line no-sync
-  await execFileSync(...execArgs);
+  if (!canSign()) {
+    debug('Skipping signing. Missing credentials.');
+    return;
+  }
+
+  const remoteServerOptions = {
+    host: process.env.SIGNING_SERVER_HOSTNAME,
+    username: process.env.SIGNING_SERVER_USERNAME,
+    port: process.env.SIGNING_SERVER_PORT,
+    privateKey: process.env.SIGNING_SERVER_PRIVATE_KEY,
+  };
+
+  if (variant === 'ubuntu') {
```

Comment:
NIT: feels that the variant here just dictates the 'local' / 'remote', and the extension is determining the method, this is not immediately understandable from the code, since we have if conditions combining both the variant and the filename.

I was thinking this code would represent reality a bit better, and be simpler with something like this:

``` js
const clientOptions = variant === 'ubuntu' ? { 
    client: 'local'
} : {
    client: 'remote',  host: process.env.SIGNING_SERVER_HOSTNAME,
    username: process.env.SIGNING_SERVER_USERNAME,
    port: process.env.SIGNING_SERVER_PORT,
    privateKey: process.env.SIGNING_SERVER_PRIVATE_KEY
}

const signingMethod = path.extname(src) === '.exe' || path.extname(src) === '.msi' ? 'jsign' : 'gpg';

return await garasign(src, {
  ...clientOptions,
  signingMethod,
});
```

Reply:
added in [d1ffa71](https://github.com/mongodb-js/compass/pull/5349/commits/d1ffa71970219314752073db1bea737a6a434951)